Support for pytree in nonstrict_traceable output

aorenste opened the pull request:
To support PyTree in nonstrict tracing we flatten the output in the proxied call and then trace the unflattening into the graph. This is similar (but reverse) to how we handle PyTree inputs into the nonstrict tracable sections.

Modified FlatApply to support unchecked (non-graphable) outputs to support this.

Added some tests for testing inputs/outputs for nonstrict traced functions.

Fixes #168890

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #168934



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @chenyang78

aorenste committed:
Support for pytree in nonstrict_traceable output

[ghstack-poisoned]

aorenste committed:
Update on "Support for pytree in nonstrict_traceable output"

Fixes #168890




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

aorenste committed:
Update on "Support for pytree in nonstrict_traceable output"

Fixes #168890




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

aorenste committed:
Update on "Support for pytree in nonstrict_traceable output"

Fixes #168890




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

aorenste committed:
Update on "Support for pytree in nonstrict_traceable output"

Fixes #168890




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1855,13 +1856,18 @@ def patched_fn(*args, **kwargs):
         # 2. Create a proxy call to `flat_apply`, then fake-tensor propagate
         # the call and wrap output into a VariableTracker.
-        proxy = tx.output.create_proxy("call_function", flat_apply, all_args, {})
+        flat_apply = FlatApply()
```
flat_apply is supposed to be a singleton, why did this change?

zou3519 reviewed (commented):
Two main things:
1) it is not clear to me why flatten_outputs needs to be an argument to `__init__`. flat_apply should be a singleton (that's how the other HOPs are)
2) can we see an expecttest for the flat_apply dataclass output?

aorenste commented on torch/_higher_order_ops/flat_apply.py:
```diff
@@ -82,14 +82,23 @@ def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _R:
-    def __init__(self) -> None:
+    # If True (default) then the output is flattened. If False then the output
+    # is returned verbatim (and must be valid graphable output).
+    flatten_output: bool = True
```
@zou3519 The reason for the change to a non-singleton was actually not from this. It needs to be an instance is for the TreeSpec. If we make it a singleton then there's no way to get out the TreeSpec for the wrapped function.  I have an idea for an alternative (which you might also not like, lol) which keeps this as a singleton. I'll make the change and update.

aorenste committed:
Update on "Support for pytree in nonstrict_traceable output"

Tweak FlatApply to support returning its result as a flattened pytree.  We need to keep it optional because to use it you need to know in advance what the output shape (TreeSpec) will be. We can use that for nonstrict tracing because we can call the function during tracing to get the TreeSpec - but there are other uses of FlatApply where we can't do that so we need to retain the prior behavior.

Added some tests for testing inputs/outputs for nonstrict traced functions.

Fixes #168890




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

aorenste committed:
Update on "Support for pytree in nonstrict_traceable output"

Tweak FlatApply to support returning its result as a flattened pytree.  We need to keep it optional because to use it you need to know in advance what the output shape (TreeSpec) will be. We can use that for nonstrict tracing because we can call the function during tracing to get the TreeSpec - but there are other uses of FlatApply where we can't do that so we need to retain the prior behavior.

Added some tests for testing inputs/outputs for nonstrict traced functions.

Fixes #168890




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1862,13 +1864,23 @@ def patched_fn(*args, **kwargs):
         # 2. Create a proxy call to `flat_apply`, then fake-tensor propagate
         # the call and wrap output into a VariableTracker.
-        proxy = tx.output.create_proxy("call_function", flat_apply, all_args, {})
+
```
@aorenste I think I understand the problem. Could you let me know if this is correct?

```
@nonstrict_trace
def g(x):
    y = x.sin()
    z = x.cos()
    return [[x], y]
    
# function being compiled
def f(x):
	result = g(x)
	return result

# desired dynamo graph
def dynamo_graph(x):
	x, y = flat_apply(g, x, flat_out=True)
	return x, y
```

The problem is:
- Dynamo's going to proxy the flat_apply call into the graph
- The output of the function `g` in Dynamo needs to be a ListVariable([ListVariable([TensorVariable]), TensorVariable])
- Dynamo usually does a FakeTensor prop and then uses the result to construct the relevant output structure.
- However if Dynamo does FakeTensorProp on this flat_apply function, it'll get [FakeTensor, FakeTensor] out of it, which unfortunately is not the right structure
- So we need some way to tell Dynamo what the real structure is.
- We need to do a FakeTensorProp to actually figure out the right structure
- So we piggyback off of Dynamo doing a FakeTensor prop (on a version of flat_apply that records the output spec!) to find the right structure

zou3519 replied:
If so can you put the example into a code comment? I think that's what helped me understand what was going on

zou3519 reviewed (commented):
let me know if I understood the problem correctly. If I did, I think I'd prefer proxying a function instead of creating a new HOP. I think we'll want to use this trick more in the future and needing to create a new HOP every time we want to do this is a bit cumbersome.

If it's too difficult to figure out how to get that to work, or if you want to punt this to a followup for someone (doesn't have to be you), please shout and I'll approve this PR

aorenste committed:
Update on "Support for pytree in nonstrict_traceable output"

Tweak FlatApply to support returning its result as a flattened pytree.  We need to keep it optional because to use it you need to know in advance what the output shape (TreeSpec) will be. We can use that for nonstrict tracing because we can call the function during tracing to get the TreeSpec - but there are other uses of FlatApply where we can't do that so we need to retain the prior behavior.

Also added FlatApplyCaptureSpec to help capture the TreeSpec to be used by FlatApply callers.

Added some tests for testing inputs/outputs for nonstrict traced functions.

Fixes #168890




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

aorenste committed:
Update on "Support for pytree in nonstrict_traceable output"

To support PyTree in nonstrict tracing we flatten the output in the proxied call and then trace the unflattening into the graph. This is similar (but reverse) to how we handle PyTree inputs into the nonstrict tracable sections.

Modified FlatApply to support unchecked (non-graphable) outputs to support this.

Added some tests for testing inputs/outputs for nonstrict traced functions.

Fixes #168890




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_flat_apply.py:
```diff
@@ -128,6 +169,7 @@ def trace_point(p):
 
         @torch._dynamo.nonstrict_trace
         def trace_point_tensor(pt):
+            # import fbvscode ; fbvscode.set_trace()
```
don't need this anymore?

zou3519 reviewed (approved):
nice, thank you

aorenste committed:
Update on "Support for pytree in nonstrict_traceable output"

To support PyTree in nonstrict tracing we flatten the output in the proxied call and then trace the unflattening into the graph. This is similar (but reverse) to how we handle PyTree inputs into the nonstrict tracable sections.

Modified FlatApply to support unchecked (non-graphable) outputs to support this.

Added some tests for testing inputs/outputs for nonstrict traced functions.

Fixes #168890




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

aorenste commented:
@pytorchbot merge
